tools/system_libs.py

```diff
@@ -574,99 +579,126 @@ class Dummy(object):
 
   if shared.Settings.USE_PTHREADS:
-    system_libs += [('pthreads',       ext, create_pthreads,       pthreads_symbols,       [libc_name],  False), # noqa
-                    ('pthreads_asmjs', ext, create_pthreads_asmjs, asmjs_pthreads_symbols, [libc_name],  False)] # noqa
+    system_libs += [Library('pthreads',       ext, create_pthreads,       pthreads_symbols,       [libc_name],  False), # noqa
+                    Library('pthreads_asmjs', ext, create_pthreads_asmjs, asmjs_pthreads_symbols, [libc_name],  False)] # noqa
 
-  system_libs += [(libc_name, ext, create_libc, libc_symbols, libc_deps, False)]
+  system_libs.append(Library(libc_name, ext, create_libc, libc_symbols, libc_deps, False))
 
   # if building to wasm, we need more math code, since we have less builtins
   if shared.Settings.WASM:
-    system_libs += [('wasm-libc', ext, create_wasm_libc, wasm_libc_symbols, [], False)]
+    system_libs.append(Library('wasm-libc', ext, create_wasm_libc, wasm_libc_symbols, [], False))
 
   # Add libc-extras at the end, as libc may end up requiring them, and they depend on nothing.
-  system_libs += [('libc-extras', ext, create_libc_extras, libc_extras_symbols, [], False)]
+  system_libs.append(Library('libc-extras', ext, create_libc_extras, libc_extras_symbols, [], False))
 
-  force.add(malloc_name())
-  if shared.Settings.WASM_BACKEND:
-    force.add('compiler-rt')
+  ret = []
+  included = set()
 
-  # Go over libraries to figure out which we must include
   def maybe_noexcept(name):
     if shared.Settings.DISABLE_EXCEPTION_CATCHING:
       name += '_noexcept'
     return name
-  ret = []
-  has = need = None
 
-  all_names = [s[0] for s in system_libs]
+  system_libs_map = {l.shortname: l for l in system_libs}
+
+  def add_library(lib):
+    if lib.shortname in included:
+      return
+    included.add(lib.shortname)
 
-  for shortname, suffix, create, library_symbols, deps, can_noexcept in system_libs:
-    assert all(d in all_names for d in deps)
-    force_this = force_all or shortname in force
-    if can_noexcept:
+    shortname = lib.shortname
+    if lib.can_noexcept:
       shortname = maybe_noexcept(shortname)
-    if force_this and not shared.Settings.WASM_OBJECT_FILES:
-      # .a files do not always link in all their parts; don't use them when forced
-      # When using wasm object files there are only .a archives but they get
-      # included via --whole-archive.
-      suffix = 'bc'
-    name = shortname + '.' + suffix
-
-    if not force_this:
-      need = set()
-      has = set()
+    name = shortname + '.' + lib.suffix
+
+    logging.debug('including %s' % name)
+
+    def do_create():
+      return lib.create(name)
+
+    libfile = shared.Cache.get(name, do_create, extension=lib.suffix)
+    need_whole_archive = lib.shortname in force and lib.suffix != 'bc'
+    ret.append((libfile, need_whole_archive))
```

Comment:
It's kinda weird that this is called ret, but is closed over. As in, we have a ret variable, that we do not return immediately. This implies the name is too short for its scope, on account of it's crossing scopes. We also no longer return ret, we return final.
libfile_pairs?